packages/Webkul/Admin/src/Services/DashboardService.php

```diff
@@ -0,0 +1,280 @@
+            return $current ? 100 : 0;
+        }
+
+        return ($current - $previous) / $previous * 100;
+    }
+
+    /**
+     * @return array
+     */
+    private function getTotalCustomers(): array
+    {
+        return [
+            'previous' => $previous = $this->customerRepository->getCustomersCountByDate($this->lastStartDate, $this->lastEndDate),
+            'current' => $current = $this->customerRepository->getCustomersCountByDate($this->startDate, $this->endDate),
+            'progress' => $this->getPercentageChange($previous, $current)
+        ];
+    }
+
+    /**
+     * @return array
+     */
+    private function getTotalOrders(): array
+    {
+        return [
+            'previous' => $previous = $this->orderRepository->getOrdersCountByDate($this->lastStartDate, $this->lastEndDate),
+            'current' => $current = $this->orderRepository->getOrdersCountByDate($this->startDate, $this->endDate),
+            'progress' => $this->getPercentageChange($previous, $current)
+        ];
+    }
+
+    /**
+     * @return array
+     */
+    private function getTotalSales(): array
+    {
+        return [
+            'previous' => $previous = $this->orderRepository->calculateSaleAmountByDate($this->lastStartDate, $this->lastEndDate),
+            'current' => $current = $this->orderRepository->calculateSaleAmountByDate($this->startDate, $this->endDate),
+            'progress' => $this->getPercentageChange($previous, $current)
+        ];
+    }
+
+    /**
+     * @return array
+     */
+    private function getAvgSales(): array
+    {
+        return [
+            'previous' => $previous = $this->orderRepository->calculateAvgSaleAmountByDate($this->lastStartDate, $this->lastEndDate),
+            'current' => $current = $this->orderRepository->calculateAvgSaleAmountByDate($this->startDate, $this->endDate),
+            'progress' => $this->getPercentageChange($previous, $current)
+        ];
+    }
+
+    /**
+     * @return float
+     */
+    private function getTotalAmountOfPendingInvoices(): float
+    {
+        return $this->invoiceRepository->getTotalAmountOfPendingInvoices();
+    }
+
+    /**
+     * @return SupportCollection
+     */
+    private function getTopSellingCategories(): SupportCollection
+    {
+        $orderItems = $this->orderItemRepository->getTopSellingOrderItemsByDate($this->startDate, $this->endDate);
+        $categories = $this->getCategoryStats($orderItems)->collapse()->unique('category_id')->values();
+
+        return $categories->map(fn (array $category) => (object) $category);
+    }
+
+    /**
+     * @param Collection $orderItems
+     * @return SupportCollection
+     */
+    private function getCategoryStats(Collection $orderItems): SupportCollection
+    {
+        return $orderItems->map(function ($orderItem) {
+            $product = $this->productRepository->find($orderItem->product_id);
+
+            return $this->getCategoryDetails($product, $orderItem);
+        });
```

Comment:
Can it be more optimized if possible its looks like it is hitting multiple queries resulting in DDoS?